File: 1283.find-the-smallest-divisor-given-a-threshold.java
/*
 * @lc app=leetcode id=1283 lang=java
 *
 * [1283] Find the Smallest Divisor Given a Threshold
 */

// @lc code=start

class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        // Find the smallest divisor such that the result mentioned above is less than
        // or equal to threshold.
        int left = 1, right = 1000000;

        while (left < right) {
            int mid = (left + right) / 2;

            if (getSum(nums, mid) <= threshold) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    public int getSum(int[] nums, int mid) {
        int sum = 0;
        for (int n : nums) {
            sum += (n - 1) / mid + 1;
        }
        return sum;
    }
}
// @lc code=end

File: 1283.find-the-smallest-divisor-given-a-threshold.py
#
# @lc app=leetcode id=1283 lang=python3
#
# [1283] Find the Smallest Divisor Given a Threshold
#

# @lc code=start

class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        left, right = 1, max(nums)
        while left < right:
            mid = (left + right) // 2
            sum = 0
            for num in nums:
                sum += math.ceil(num/mid)
            if sum <= threshold:
                right = mid
            else:
                left = mid + 1
        return left


# @lc code=end
